salt/modules/virt.py

```diff
@@ -599,18 +601,29 @@ def _gen_xml(
                 context["boot"]["kernel"] = "/usr/lib/grub2/x86_64-xen/grub.xen"
                 context["boot_dev"] = []
 
-    if "serial_type" in kwargs:
-        context["serial_type"] = kwargs["serial_type"]
-    if "serial_type" in context and context["serial_type"] == "tcp":
-        if "telnet_port" in kwargs:
-            context["telnet_port"] = kwargs["telnet_port"]
-        else:
-            context["telnet_port"] = 23023  # FIXME: use random unused port
-    if "serial_type" in context:
-        if "console" in kwargs:
-            context["console"] = kwargs["console"]
-        else:
-            context["console"] = True
+    context["serials"] = []
+    for serial in enumerate(serials):
+        serial_context = {
+            "type": serial.type
+        }
```

Comment:
You need to take into account the old way to give the type, use the above mentioned `default_serial_type` too

Also don't assume the `type` property will always be in the `serial` dict... users may forget it and it would explode here. You should guard the whole loop with something like:

```suggestion
        serial_type = serial.get("type") or default_serial_type
        if not serial_type:
            throw SaltInvocationError("Missing type in serial")
            
        serial_context = {
            "type": serial_type
        }
```